src/manifolds/Tucker.jl

```diff
@@ -0,0 +1,741 @@
+            return DomainError(
+                norm(U' * U̇),
+                s *
+                "since the columns of x.hosvd.U are not" *
+                "orthogonal to those of v.U̇.",
+            )
+        end
+    end
+    return nothing
+end
+
+"""
+    Base.convert(::Type{Matrix{T}}, basis :: CachedBasis{𝔽,DefaultOrthonormalBasis{𝔽, TangentSpaceType},HOSVDBasis{T, D}}) where {𝔽, T, D}
+    Base.convert(::Type{Matrix}, basis :: CachedBasis{𝔽,DefaultOrthonormalBasis{𝔽, TangentSpaceType},HOSVDBasis{T, D}}) where {𝔽, T, D}
+
+Convert a HOSVD basis to a matrix whose columns are the vectorisations of the basis vectors.
+"""
+function Base.convert(::Type{Matrix{T}}, ℬ::CachedHOSVDBasis{𝔽,T,D}) where {𝔽,T,D}
+    𝔄 = ℬ.data.point
+    r⃗ = size(𝔄.hosvd.core)
+    n⃗ = size(𝔄)
+    ℳ = Tucker(n⃗, r⃗)
+
+    J = Matrix{T}(undef, prod(n⃗), manifold_dimension(ℳ))
+    # compute all possible ∂𝔄╱∂ℭ (in one go is quicker than one vector at a time)
+    J[:, 1:prod(r⃗)] = ⊗ᴿ(𝔄.hosvd.U...)
+    # compute all possible ∂𝔄╱∂U[d] for d = 1,...,D
+    function fill_column!(i, vᵢ)
+        Jᵢ_tensor = reshape(view(J, :, i), n⃗) # changes to this apply to J as well
+        return embed!(ℳ, Jᵢ_tensor, 𝔄, vᵢ)
+    end
+    foreach(fill_column!, ℳ, 𝔄, ℬ, (prod(r⃗) + 1):manifold_dimension(ℳ))
+    return J
+end
+function Base.convert(::Type{Matrix}, basis::CachedHOSVDBasis{𝔽,T,D}) where {𝔽,T,D}
+    return convert(Matrix{T}, basis)
+end
+
+Base.copy(x::TuckerTVector) = TuckerTVector(copy(x.Ċ), map(copy, x.U̇))
+
+function Base.copyto!(q::TuckerPoint, p::TuckerPoint)
+    for d in 1:ndims(q)
+        copyto!(q.hosvd.U[d], p.hosvd.U[d])
+        copyto!(q.hosvd.σ[d], p.hosvd.σ[d])
+    end
+    copyto!(q.hosvd.core, p.hosvd.core)
+    return q
+end
+function Base.copyto!(y::TuckerTVector, x::TuckerTVector)
+    for d in 1:ndims(y.Ċ)
+        copyto!(y.U̇[d], x.U̇[d])
+    end
+    copyto!(y.Ċ, x.Ċ)
+    return y
+end
+
+@doc raw"""
+    embed(::Tucker, A :: TuckerPoint)
+
+Convert a point `A` on the Tucker manifold to a full tensor, represented as an
+$N_1 \times \dots \times N_D$-array.
+
+    embed(::Tucker, A::TuckerPoint, X::TuckerTVector)
+
+Convert a tangent vector `X` with base point `A` on the Tucker manifold to a full tensor,
+epresented as an $N_1 \times \dots \times N_D$-array.
+"""
+embed(::Tucker, ::Any, ::TuckerPoint)
+
+function embed!(::Tucker, q, p::TuckerPoint)
+    return copyto!(q, reshape(⊗ᴿ(p.hosvd.U...) * vec(p.hosvd.core), size(p)))
+end
+function embed!(ℳ::Tucker, Y, 𝔄::TuckerPoint{T,D}, X::TuckerTVector) where {T,D}
+    Y .= reshape(⊗ᴿ(𝔄.hosvd.U...) * vec(X.Ċ), size(Y))
+    Uℭ = embed(ℳ, 𝔄)
+    n⃗ = size(Uℭ)
+    for d in 1:D
+        Y .= Y + fold(X.U̇[d] * (𝔄.hosvd.U[d]' * unfold(Uℭ, d)), d, n⃗)
+    end
+    return Y
+end
+
+# Inverse of the k'th unfolding of a size n₁ × ... × n_D tensor
+function fold(𝔄♭::AbstractMatrix{T}, k, n⃗::NTuple{D,Int})::Array{T,D} where {T,D,Int}
```

Comment:
Could you change this name to, for example, `tucker_fold`? This way it won't be confused with any other folding.